src/trusted/service_runtime/sel_ldr.c

```diff
@@ -1619,27 +1616,34 @@ void NaClCopyDynamicTextAndVmmap(struct NaClApp *nap_parent, struct NaClApp *nap
       struct NaClVmmapEntry *entry = parentmap->vmentry[i];
       uintptr_t page_addr_child = (entry->page_num << NACL_PAGESHIFT) | offset;
       size_t copy_size = entry->npages << NACL_PAGESHIFT;
-      NaClVmmapAddWithOverwrite(&nap_child->mem_map,
-                                entry->page_num,
-                                entry->npages,
-                                entry->prot,
-                                (entry->flags | MAP_ANON_PRIV) & ~NACL_ABI_MAP_SHARED,
-                                entry->desc,
-                                entry->offset,
-                                entry->file_size);
-
-  
-      if(entry->prot && (entry->desc != nap_parent->text_shm)) {
-        size_t endaddr = page_addr_child + copy_size;
-        if (NaClMprotect((void *)page_addr_child, copy_size, entry->prot) == -1) {
-          NaClLog(LOG_FATAL, "%s\n", "parent vmmap page NaClMprotect failed!");
+      if(entry->desc != nap_parent->text_shm) {
+        struct NaClDesc* desc = entry->desc;
+        if(entry->desc) {
+          struct NaClHostDesc* hd = ((struct NaClDescIoDesc*) desc)->hd;
+          desc = NaClGetDesc(nap_child, fd_cage_table[nap_child->cage_id][hd->lindfd]);
         }
-        if(endaddr > child_stack_addr && child_stack_addr > page_addr_child) {
-          //likwise, if the mapping is the stack, only patch up to the stack pointer
-          NaClPatchAddr(offset, parent_offset, (uintptr_t *)child_stack_addr, endaddr - child_stack_addr);
-        } else {
-          NaClPatchAddr(offset, parent_offset, (uintptr_t *)page_addr_child, copy_size);
+        NaClVmmapAddWithOverwrite(&nap_child->mem_map,
+                                  entry->page_num,
+                                  entry->npages,
+                                  entry->prot,
+                                  (entry->flags | MAP_ANON_PRIV) & ~NACL_ABI_MAP_SHARED,
+                                  desc,
+                                  entry->offset,
+                                  entry->file_size);
+        if(entry->prot) {
+          if (NaClMprotect((void *)page_addr_child, copy_size, entry->prot) == -1) {
+            NaClLog(LOG_FATAL, "%s\n", "parent vmmap page NaClMprotect failed!");
+          }
         }
+      } else {
+        NaClVmmapAddWithOverwrite(&nap_child->mem_map,
+                                  entry->page_num,
+                                  entry->npages,
+                                  entry->prot,
+                                  (entry->flags | MAP_ANON_PRIV) & ~NACL_ABI_MAP_SHARED,
+                                  entry->desc,
```

Comment:
Is this supposed to be the child desc?

Reply:
Nope, this one is just for the text_shm desc, which is a Shm desc and is not an io desc, and therefore has no associated fd, and also no associated cage, so nothing needs to be changed about which desc is associated on fork.